### ISSUE TITLE
[BUG] Change in behaviour when initializing a fused memoryview with `None`

### ISSUE THREAD

**Issue by da-woods** [cython/cython#5297]
### Describe the bug

In Cython 0.29.x initializing a fused memoryview with `None` picked a type (probably just the first on the list).

In Cython 3.0.0b1. initializing a fused memoryview with `None` fails.

### Code to reproduce the behaviour:

```cython
from cython cimport floating, typeof

def f(floating[::1] d=None):
    if d is not None:
        return d[0]
    else:
        return typeof(d)

print(f())
```

### Expected behaviour

In Cython 0.29.x this prints 'float' while in Cython 3.0 raises `TypeError: No matching signature found`.

Personally I'm not sure what's "right". But for the moment it's a regression. Cython 0.29.x is kind of buggy in its own way:

```
from cython cimport floating, typeof

def f(floating[::1] d, floating[::1] e):
    if d is not None:
        return d[0]
    else:
        return typeof(d)

import numpy as np
arr = np.zeros((1,), dtype=np.float64)
print(f(None, arr))
```

It selects `float32` on the basis that this matches argument `d`, and then errors with `Buffer dtype mismatch, expected 'float' but got 'double'` on argument `e`.

### Environment

Cython 3.0.0b1

### Additional context

Bisected to 805d7f3d778cf29f1e2acaaac84cd2bcadb25e14 - I don't think it'd be too hard to restore the previous behaviour, but I'm not sure the previous behaviour is good.

Reported https://groups.google.com/g/cython-users/c/dVomEXqJgX8

**Comment by scoder** [comment]
Yes, both kinds of behaviour seem somewhat wrong, intransparent or unhelpful.

I considered an explicit type cast for the `None` default argument, but that wouldn't work with multiple arguments (including unconstrained non-default ones) since you may still end up with inconsistent types then. Let's look at the different cases.

1. `func(floating[:] x = None)` – a type cast for `None` to a specific memory view type (e.g. `<float[:]>None`) would help to select the right default case.
2. `func(floating[:] a, floating[:] b = None)` – excluding `b` from the list of arguments for the default case would allow to select a good variant. An explicit cast for `None` (e.g. `<float[:]>None`) would lead to call time conflicts with the other type case(s).
3. `func(floating[:] a, integrals[:] b = None)` – an explicit type cast would help.
4. `func(floating[:] a, integrals[:] b = None, integrals[:] c = None)` – an explicit type cast would help but is only required for `b` or `c`, not both. In any case, the type casts much not conflict.
5. `func(floating[:] a, floating[:] b = None, integrals[:] c = None)` – an explicit type cast is needed only for `c`.

Basically, whenever the default case is required to select the right variant, we need a user cast and should at least warn about the ambiguity.

When other non-default arguments of the same fused type are provided, the default argument should best be excluded from the type selection vector.

Since type casts in Python syntax are verbose, this gets somewhat ugly in Python code:
`func(a: floating[:], b: floating[:] = cython.cast(cython.float[:], None))`

**Comment by da-woods** [comment]
I don't think type-casts can be the answer. These are `def` functions that are called as normal Python objects (and likely from Python) so I don't think the type-cast can be transmitted to the call.

The example I picked might be a little misleading: this isn't *just* about default arguments. You could have a signature of `f(floating[:] a, floating[:] b)` and call it as `f(None, some_array)`. Which adds more options than you've shown above.

The current behaviour is "deduce the each fused type from the left-most argument it appears in"

Looking at your individual examples:

1. `func(floating[:] x = None)` can't really unambiguously deduce `x`. The choice is fail, or pick something arbitrary maybe with a warning. Picking something arbitrary preserves compatibility.
2. `func(floating[:] a, floating[:] b = None)` - current behaviour just uses the type of `a` to deduce. That's remains fine.
3. `func(floating[:] a, integrals[:] b = None)` - same as "1"
4. `func(floating[:] a, integrals[:] b = None, integrals[:] c = None)` same as "1"
5. `func(floating[:] a, floating[:] b = None, integrals[:] c = None)` same as "1"

Then the additional `func(None, some_array)` case - current behaviour is pick something arbitrary. I'd propose it should work it out from `b` if possible.

What I'd suggest is:
* For each fused type, keep going right until it can work something out unambiguously. That complicates things a little but not too much since it still only considers one argument at a time.
* If a type can't be worked out unambiguously then pick the first fused type and emit a runtime warning. My reasoning is:
  * Backwards compatibility
  * Mostly it doesn't matter - all of the overloads will work the same so picking the first one is reasonable.

It's a medium-sized change to the fused type dispatch code. #5298 may still be worthwhile as an interim measure to bring things back to where they were. But I don't think it's the long-term solution.

**Comment by scoder** [comment]
> These are `def` functions that are called as normal Python objects (and likely from Python) so I don't think the type-cast can be transmitted to the call.

I meant a type cast for the default argument, as part of the signature definition. That would help in some cases, as I described. Casting a call argument would not help with calling def functions, right, but it would still help with fused cdef/cpdef functions.

**Comment by da-woods** [comment]
> > These are `def` functions that are called as normal Python objects (and likely from Python) so I don't think the type-cast can be transmitted to the call.
> 
> I meant a type cast for the default argument, as part of the signature definition. That would help in some cases, as I described. Casting a `None` call argument would not help with calling def functions, right, but it would still help with fused cdef/cpdef functions.

Thanks for the clarification. I misunderstood. That could help to set a default type where no type is given
